module/documents/actor/actor.mjs

```diff
@@ -799,38 +788,41 @@ export default class Actor5e extends SystemDocumentMixin(Actor) {
     let keyLevel = Math.clamp(progression[key], 0, CONFIG.DND5E.maxLevel);
     spells[key] ??= {};
     spells[key].type = key;
+    spells[key].label = game.i18n.localize(CONFIG.DND5E.spellcasting[key]?.label ?? "");
     const override = Number.isNumeric(spells[key].override) ? parseInt(spells[key].override) : null;
 
     // Slot override
     if ( (keyLevel === 0) && (actor.type === "npc") && (override !== null) ) {
       keyLevel = actor.system.details.spellLevel;
     }
 
-    const [, keyConfig] = Object.entries(table).reverse().find(([l]) => Number(l) <= keyLevel) ?? [];
-    if ( keyConfig ) {
-      spells[key].level = keyConfig.level;
-      if ( override === null ) spells[key].max = keyConfig.slots;
-      else spells[key].max = Math.max(override, 1);
-      spells[key].value = Math.min(spells[key].value, spells[key].max);
-    }
-
-    else {
-      spells[key].max = override || 0;
-      spells[key].level = (spells[key].max > 0) ? 1 : 0;
-    }
+    // Apply slots
+    const configuration = CONFIG.DND5E.spellcasting.calculateSlots(key, keyLevel);
+    if ( foundry.utils.isEmpty(configuration) && !override ) return;
+    const [[level, amount]=[]] = Object.entries(configuration);
+    spells[key].max = Number.isNumeric(override) ? Math.max(parseInt(override), 0) : (amount || 0);
+    spells[key].level = (spells[key].max > 0) ? (level ? parseInt(level) : 1) : 0;
+    spells[key].value = Math.clamp(spells[key].value, 0, spells[key].max);
   }
 
   /* -------------------------------------------- */
 
   /**
-   * Convenience method for preparing pact slots specifically.
-   * @param {object} spells        The `data.spells` object within actor's data. *Will be mutated.*
+   * Prepare alternative leveled spell slots using progression data.
+   * @param {object} spells        The `system.spells` object within actor's data. *Will be mutated.*
    * @param {Actor5e} actor        Actor for whom the data is being prepared.
    * @param {object} progression   Spellcasting progression data.
+   * @param {string} key           The internal key for these spell slots on the actor.
    */
-  static preparePactSlots(spells, actor, progression) {
-    this.prepareAltSlots(spells, actor, progression, "pact", CONFIG.DND5E.pactCastingProgression);
-    spells.pact.label = game.i18n.localize("DND5E.PactMagic");
+  static prepareLeveledSlots(spells, actor, progression, key) {
+    const slots = CONFIG.DND5E.spellcasting.calculateSlots(key, progression[key]);
+    for ( const level of Array.fromRange(Object.keys(CONFIG.DND5E.spellLevels).length - 1, 1) ) {
+      const slot = spells[`${key}${level}`] ??= { value: 0 };
```

Comment:
During data prep, if the spellcasting is `separate`, then objects of with a key of `<name><level>` will be created (such as "spell1" through "spell9"), otherwise a single object is created (eg "pact").

This is entirely dynamic, and can be tested by setting `pact.table = "spell"` and `pact.separate = true`, which results in `pact1` through `pact9`.

![image](https://github.com/user-attachments/assets/ef908dfd-1463-4177-862d-74fffc9942a6)